In [2]:
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split
import os

In [3]:
project_dir="/Users/karthik/Desktop/SAGEMAKER"
data_dir="DATASET"

In [4]:
def get_data(name):
    name=str(name)
    file_name=f"{name}.csv"
    file_path=os.path.join(project_dir,data_dir,file_name)
    return pd.read_csv(file_path)

In [5]:
flight=get_data("flight_price")

In [6]:
flight

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [7]:
flight.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


THE DATASET CONTAINS 10,683 COLUMNS, 

10 FEATURES,

ROUTE AND TOTAL_STOP ARE MISSING 1 VALUE EACH

## CHECK DATA TYPE


In [8]:
flight.head(2)

Airline Date_of_Journey    Source Destination                  Route  \
0     IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1  Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662

In [9]:
flight[flight.duplicated(keep=False)].sort_values(["Airline","Date_of_Journey","Source","Destination"])

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

## OBSERVATION


"total stops" can be numberic

duration can be converted into int type

date of journey should be converted to date Time

Dep Time as date time 

arrival time as date time 


# DETAILED ANALYSIS

In [10]:
def convert_to_minutes(ser):
    return(
        ser
        .str.split(" ",expand=True)
        .set_axis(["hour","minute"],axis=1)
        .assign(
            hour=lambda df_:(
                df_
                .hour
                .str.replace("h","")
                .astype(int)
                .mul(60)
            ),
            minute=lambda df_:(
            df_
            .minute
            .str.replace("m","")
            .fillna(0)
            .astype(int)
            ))
            .sum(axis=1)
    )
        
    

In [46]:
def clean_data(df):
    return (
        df
        .drop(index=6474)
        .drop_duplicates()
        .assign(**{
            col:df[col].str.strip()
            for col in df.select_dtypes(include="O").columns})
        
        .rename(columns=str.lower)
        
        .assign(
            airline=lambda df_:df_.airline.str.lower())
        
        .assign(
            airline=lambda df_:
                (df_
                .airline
                .str.replace(" Premium economy","")
                .str.replace(" Business","")
                .str.title()),
            
            date_of_journey= lambda df_:
            pd.to_datetime(df_.date_of_journey,dayfirst=True),
            
            dep_time=lambda df_:pd.to_datetime(df_.dep_time,).dt.time,
            
            arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,
            
            duration=lambda df_:df_.duration.pipe(convert_to_minutes),
            
            total_stops=lambda df_:df_ .total_stops
                .replace("non-stop","0")
                .str.replace(" stops?","",regex=True)
                .pipe(lambda ser:pd.to_numeric(ser)),

            additional_info=lambda df_:df_.additional_info.replace("No info","No Info")
            )
        .drop("route",axis=1)
        
    )

In [47]:
clean_data(flight).info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   airline          10462 non-null  object        
 1   date_of_journey  10462 non-null  datetime64[ns]
 2   source           10462 non-null  object        
 3   destination      10462 non-null  object        
 4   dep_time         10462 non-null  object        
 5   arrival_time     10462 non-null  object        
 6   duration         10462 non-null  int64         
 7   total_stops      10461 non-null  float64       
 8   additional_info  10462 non-null  object        
 9   price            10462 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(2), object(6)
memory usage: 899.1+ KB


/var/folders/d3/txrppy7x21n69k6039v1wbwh0000gn/T/ipykernel_33258/3504186817.py:26: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_:pd.to_datetime(df_.dep_time,).dt.time,
/var/folders/d3/txrppy7x21n69k6039v1wbwh0000gn/T/ipykernel_33258/3504186817.py:28: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,


In [13]:
clean_data(flight).info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   airline          10462 non-null  object        
 1   date_of_journey  10462 non-null  datetime64[ns]
 2   source           10462 non-null  object        
 3   destination      10462 non-null  object        
 4   dep_time         10462 non-null  object        
 5   arrival_time     10462 non-null  object        
 6   duration         10462 non-null  int64         
 7   total_stops      10461 non-null  float64       
 8   additional_info  10462 non-null  object        
 9   price            10462 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(2), object(6)
memory usage: 899.1+ KB


/var/folders/d3/txrppy7x21n69k6039v1wbwh0000gn/T/ipykernel_33258/3504186817.py:26: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_:pd.to_datetime(df_.dep_time,).dt.time,
/var/folders/d3/txrppy7x21n69k6039v1wbwh0000gn/T/ipykernel_33258/3504186817.py:28: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,


some entries have premium nd buisness in them


In [14]:
flight.head(1)

Airline Date_of_Journey    Source Destination      Route Dep_Time  \
0  IndiGo      24/03/2019  Banglore   New Delhi  BLR → DEL    22:20   

   Arrival_Time Duration Total_Stops Additional_Info  Price  
0  01:10 22 Mar   2h 50m    non-stop         No info   3897

In [15]:
(
    flight
    .Total_Stops
    .replace("non-stop","0")
    .str.replace(" stops?","",regex=True)
    .pipe(lambda ser:pd.to_numeric(ser))
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

# split dataset

In [17]:
fight_final=clean_data(flight).sample(1000)

/var/folders/d3/txrppy7x21n69k6039v1wbwh0000gn/T/ipykernel_33258/3504186817.py:26: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_:pd.to_datetime(df_.dep_time,).dt.time,
/var/folders/d3/txrppy7x21n69k6039v1wbwh0000gn/T/ipykernel_33258/3504186817.py:28: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_:pd.to_datetime(df_.arrival_time).dt.time,


In [18]:
X=fight_final.drop("price",axis=1)
Y=fight_final.price.copy()

In [19]:
from sklearn.model_selection import train_test_split

In [34]:
X_,X_test,Y_,Y_test=train_test_split(X,Y,test_size=0.2,random_state=69)
X_train,X_val,Y_train,Y_val=train_test_split(X_,Y_,test_size=0.2,random_state=69)

In [35]:
def export_data(X,Y,name):
    file_name=f"{name}.csv"
    file_path=os.path.join(project_dir,data_dir,file_name)

    X.join(Y).to_csv(file_path,index=False)

    return pd.read_csv(file_path).head()


In [36]:
export_data(X_train,Y_train,"train")

airline date_of_journey    source destination  dep_time arrival_time  \
0       Indigo      2019-06-24  Banglore       Delhi  13:00:00     15:50:00   
1  Jet Airways      2019-05-21   Kolkata    Banglore  08:25:00     19:50:00   
2     Spicejet      2019-05-27    Mumbai   Hyderabad  22:45:00     00:15:00   
3    Air India      2019-06-06   Kolkata    Banglore  08:20:00     20:25:00   
4  Jet Airways      2019-05-15   Kolkata    Banglore  20:00:00     08:15:00   

   duration  total_stops              additional_info  price  
0       170          0.0                      No Info   3943  
1       685          1.0  In-flight meal not included  10844  
2        90          0.0                      No Info   2017  
3       725          2.0                      No Info  11432  
4       735          1.0  In-flight meal not included   8586

In [37]:
export_data(X_val,Y_val,"val")

airline date_of_journey   source destination  dep_time  \
0             Indigo      2019-05-27   Mumbai   Hyderabad  21:20:00   
1          Air India      2019-03-24  Kolkata    Banglore  05:50:00   
2             Indigo      2019-03-12  Chennai     Kolkata  22:05:00   
3        Jet Airways      2019-03-06    Delhi      Cochin  16:00:00   
4  Multiple Carriers      2019-05-21    Delhi      Cochin  10:20:00   

  arrival_time  duration  total_stops additional_info  price  
0     22:50:00        90          0.0         No Info   2227  
1     13:40:00       470          2.0         No Info   8005  
2     00:30:00       145          0.0         No Info   6297  
3     19:45:00      1665          1.0         No Info  17024  
4     19:15:00       535          1.0         No Info   8266

In [38]:
export_data(X_test,Y_test,"test")

airline date_of_journey    source destination  dep_time  \
0             Indigo      2019-03-15    Mumbai   Hyderabad  02:30:00   
1          Air India      2019-05-12   Kolkata    Banglore  10:00:00   
2        Jet Airways      2019-03-21  Banglore   New Delhi  22:50:00   
3  Multiple Carriers      2019-06-24     Delhi      Cochin  11:40:00   
4          Air India      2019-05-09     Delhi      Cochin  19:45:00   

  arrival_time  duration  total_stops              additional_info  price  
0     04:00:00        90          0.0                      No Info   3342  
1     01:20:00       920          2.0                      No Info  13243  
2     07:40:00       530          1.0  In-flight meal not included   7045  
3     01:30:00       830          1.0  In-flight meal not included   5797  
4     19:15:00      1410          2.0                      No Info  10178

In [40]:
fight_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 1298 to 9228
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   airline          1000 non-null   object        
 1   date_of_journey  1000 non-null   datetime64[ns]
 2   source           1000 non-null   object        
 3   destination      1000 non-null   object        
 4   dep_time         1000 non-null   object        
 5   arrival_time     1000 non-null   object        
 6   duration         1000 non-null   int64         
 7   total_stops      1000 non-null   float64       
 8   additional_info  1000 non-null   object        
 9   price            1000 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int64(2), object(6)
memory usage: 85.9+ KB
